packages/components/src/DataList/components/DataListLayoutInternal/DataListHeaderCheckbox.tsx

```diff
@@ -0,0 +1,72 @@
+import classNames from "classnames";
+import React from "react";
+import { ReactElement } from "react-markdown/lib/react-markdown";
+import { AnimatedSwitcher } from "../../../AnimatedSwitcher";
+import { Button } from "../../../Button";
+import { Checkbox } from "../../../Checkbox";
+import { Text } from "../../../Text";
+import { useDataListContext } from "../../context/DataListContext";
+import styles from "../../DataList.css";
+
+interface DataListHeaderCheckbox {
+  readonly children: ReactElement;
+}
+
+export function DataListHeaderCheckbox({ children }: DataListHeaderCheckbox) {
+  const {
+    data,
+    totalCount,
+    selected = [],
+    onSelectAll,
+    onSelect,
+  } = useDataListContext();
+
+  if (!onSelectAll && !onSelect) return children;
+
+  return (
+    <div className={classNames(styles.selectable)}>
+      <div
+        className={classNames(styles.selectAllCheckbox, {
+          [styles.visible]: Boolean(onSelectAll),
+        })}
+      >
+        <Checkbox
+          checked={isAllSelected()}
+          indeterminate={selected.length > 0 && !isAllSelected()}
+          onChange={onSelectAll}
+        />
+      </div>
+
+      <AnimatedSwitcher
+        switched={Boolean(selected.length)}
+        initialChild={children}
+        switchTo={
+          <div className={styles.headerBatchSelect}>
+            <Text>
+              <b>{selected.length} selected</b>
+            </Text>
+            <Button
+              label="Deselect All"
+              onClick={() => onSelect?.([])}
```

Comment:
still debating if there should be an `onDeselectAll` prop that returns `[]` always and forever

Reply:
Good question.  I think it being called onSelect still implies that it can handle selecting and deselecting - you are selecting all either way.

Reply:
I like that reason. I'll keep it as is!